### PR TITLE
Master wowl lpe

### PR DESCRIPTION
**Pull request by kebeclibre**

### CODE REVIEW

File: addons/web/static/src/fields/email_field.xml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
     <t t-name="web.EmailField" owl="1">
         <t t-if="props.readonly">
-            <a class="o_form_uri o_text_overflow" t-att-href="'mailto:'+props.value" t-esc="props.value"/>
+            <a class="o_form_uri o_text_overflow" t-att-href="'mailto:'+props.value" t-esc="props.value || ''"/>
```
Comment (aab-odoo) [inline addons/web/static/src/fields/email_field.xml]: t-att-href if value is falsy?

File: addons/web/static/src/fields/field.js [hunk 2]
```diff
@@ -13,6 +13,26 @@ const { Component, xml } = owl;
 
 const viewRegistry = registry.category("views");
 
+export function fieldVisualFeedback(record, fieldName) {
+    const cls = record.activeFields[fieldName].FieldComponent;
```
Comment (aab-odoo) [inline addons/web/static/src/fields/field.js]: should begin with an uppercase

File: addons/web/static/src/fields/integer_field.js [hunk 3]
```diff
@@ -41,7 +41,7 @@ IntegerField.defaultProps = {
     inputType: "text",
     setAsInvalid: () => {},
 };
-IntegerField.isEmpty = () => false;
+IntegerField.isEmpty = (record, fieldName) => record.data[fieldName] === false ? true : false;
```
Comment (aab-odoo) [inline addons/web/static/src/fields/integer_field.js:44]: Don't we default numeric field values to 0?

File: addons/web/static/src/views/form/button_box/button_box.js [hunk 4]
```diff
@@ -1,77 +1,32 @@
...
+            if (!("isVisible" in slot) || slot.isVisible) {
+                if (visible.length >= maxVisibleButtons) {
+                    additional.push(slotName);
+                } else {
+                    visible.push(slotName);
+                }
+
+            }
+        }
+        return {visible, additional};
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/button_box/button_box.js:28]: I remember commenting this code, sorry :D

File: addons/web/static/src/views/form/form_group/form_group.xml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      <br t-if="item.newline" />
+      <t t-slot="{{ item.name }}" className="'o_group_col_' + item.size"/>
+    </t>
+  </div>
+</t>
+
+<t t-name="web.Form.InnerGroup" owl="1">
+  <table t-attf-class="{{ allClasses }}" class="o_inner_group o_group" t-if="props.slots" t-att-style="props.style">
+      <tbody>
+          <td t-if="props.slots.title" t-attf-colspan="{{ props.maxCols }}" style="width: 100%">
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_group/form_group.xml:17]: inline style?

File: addons/web/static/src/views/relational_model.js [hunk 6]
```diff
@@ -329,7 +329,7 @@ class DataPoint {
     _parseServerValue(field, value) {
         switch (field.type) {
             case "char": {
-                return value || "";
+                return value;
```
Comment (aab-odoo) [inline addons/web/static/src/views/relational_model.js]: remove the case then :D I was pretty sure we had to do that (and to default numeric values to 0). I must be mistaken

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 7]
```diff
@@ -1772,6 +1772,10 @@ QUnit.module("Views", (hooks) => {
                 trululu(record) {
                     // when trululu changes, push another record in product_ids.
                     // only push a second record once.
+
+                    // The new relational model doesn't send un unchanged field
+                    // LPE To check
+                    record.product_ids = record.product_ids || [];
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js]: please don't do such things, we'll almost certainly not re-read this line again, and we'll merge like this. If there something strange with the test, keep it skipped, or investigate now.

**Review by aab-odoo — APPROVED** [review]
Cool PR, makes a bunch of tests pass :-)

File: addons/web/static/src/views/form/form_group/form_group.xml (hunk 5 above)
Comment (kebeclibre) [reply addons/web/static/src/views/form/form_group/form_group.xml:17]: some tests assert that, I did not want to change it everywhere... but I agree, it should directly be an attr on the node.
That is computed in JS, so I guess it is rather hard to do it in css

File: addons/web/static/src/fields/email_field.xml (hunk 1 above)
Comment (kebeclibre) [reply addons/web/static/src/fields/email_field.xml]: yes, from the model point of view, false is a legitimate value, meaning "unset".
Domains need to be able to make the difference between unset and empty string.
But, at display time, false.toString gives "false", and we want the value to be ""

File: addons/web/static/src/fields/field.js (hunk 2 above)
Comment (kebeclibre) [reply addons/web/static/src/fields/field.js]: for a function ?

File: addons/web/static/src/fields/integer_field.js (hunk 3 above)
Comment (kebeclibre) [reply addons/web/static/src/fields/integer_field.js:44]: yes, but samething as char, a numeric value can be unset

**Pull request merged**